Gems/Terrain/Assets/Shaders/Terrain/ClipmapComputeHelpers.azsli

```diff
@@ -66,13 +82,68 @@ float2 GetCurrentWorldPosition(uint2 pixelPosition, uint clipmapLevel)
 float2 ddxPosition(uint clipmapLevel)
 {
     float clipmapScale = GetClipmapScale(clipmapLevel);
-    float2 dxdy = PassSrg::m_clipmapData.m_maxRenderSize / PassSrg::m_clipmapData.m_clipmapSize * clipmapScale;
+    float2 dxdy = TerrainSrg::m_clipmapData.m_maxRenderSize / TerrainSrg::m_clipmapData.m_clipmapSize * clipmapScale;
     return float2(dxdy.x, 0.0);
 }
 
 float2 ddyPosition(uint clipmapLevel)
 {
     float clipmapScale = GetClipmapScale(clipmapLevel);
-    float2 dxdy = PassSrg::m_clipmapData.m_maxRenderSize / PassSrg::m_clipmapData.m_clipmapSize * clipmapScale;
+    float2 dxdy = TerrainSrg::m_clipmapData.m_maxRenderSize / TerrainSrg::m_clipmapData.m_clipmapSize * clipmapScale;
     return float2(0.0, dxdy.y);
 }
+
+struct ClipmapSample
+{
+    float3 m_macroColor;
+    float3 m_macroNormal;
+    bool   m_hasDetail;
+    float3 m_detailColor;
+    float3 m_detailNormal;
+    float  m_detailRoughness;
+    float  m_detailSpecularF0;
+    float  m_detailMetalness;
+    float  m_detailOcclusion;
+    float  m_height;
+};
+
+ClipmapSample SampleClipmap(float2 worldPosition)
+{
+    float2 currentViewPosition = TerrainSrg::m_clipmapData.m_viewPosition.zw;
+    float2 halfMaxRenderSize = TerrainSrg::m_clipmapData.m_maxRenderSize / 2.0;
+    // The top clipmap's render distance
+    float2 halfMinRenderSize = halfMaxRenderSize / (float)(1u << (ClipmapStackSize - 1u));
+
+    float2 distance = worldPosition - currentViewPosition;
+    float2 clampedDistance = clamp(halfMinRenderSize, abs(distance), halfMaxRenderSize);
+    float2 distanceRatio = halfMaxRenderSize / clampedDistance;
+    // Due to clamping, the max result of log2 is (ClipmapStackSize - 1u), which falls into the first clipmap
+    uint clipmapLevel = ClipmapStackSize - 1u - uint(floor(log2(min(distanceRatio.x, distanceRatio.y))));
+    uint2 pixelPosition = GetPixelPosition(distance, clipmapLevel);
+    uint3 sampleIndex = uint3(pixelPosition, clipmapLevel);
+
+    float4 macroColor = TerrainSrg::m_macroColorClipmaps[sampleIndex];
+    float4 macroNormal = TerrainSrg::m_macroNormalClipmaps[sampleIndex];
+    float4 detailColor = TerrainSrg::m_detailColorClipmaps[sampleIndex];
+
+    ClipmapSample data;
+    data.m_macroColor = macroColor.rgb;
+    data.m_macroNormal = macroNormal.xyz;
+    data.m_hasDetail = detailColor.a > 0.5;
```

Comment:
Why "a > 0.5"? Please add a comment explaining this. I would've thought an alpha of 0.1 would still count as having detail?

Reply:
I'll make a comment. I am encoding 0.0 as no and 1.0 as yes, so using 0.5 to check.

Reply:
I would suggest `detailColor.a == 1.0` with a comment then, since you are writing "data" into this channel